### DISCUSSION TITLE
Get error response object

### DISCUSSION THREAD

**Discussion by fiskpatte** [TanStack/query#1463] [General] [answered]
Hi.
A mutation I'm calling fails with 400, and in devtools I see the response:
`{"error":{"code":"INVALID_INPUT","message":"An application must have at least one redirect uri","data":null}}`
However, when I access the `error.message` from the useMutation hook, the value is just "Request failed with status code 400". How do I access the response error text in useMutation?

Sorry for this noob question but I can't figure it out.

**Comment by TkDodo** [comment, marked as answer]
It depends on what you are using for data fetching, react-query is completely agnostic in that regard.

- If you are using `axios`, it will give you an `AxiosError` which has the json body somewhere in `error.response.data` or so.
- If you are using `fetch`, you have to json parse the body yourselves via `error.response.json()`.
- If you are using something else, I don't know, please consult the docs of your fetching library :)

**Reply by bfunc** [reply]
```
 const mutationCreate = useMutation(createUser, {
    onSuccess: () => queryClient.invalidateQueries('users'),
    /* 
      error - parameter is instance of Error with message: Request failed with status code 400
      QUESTION: How to get response payload json data here?
     */
    onError: error => setError(error.message),
  })
```

**Reply by znareak** [reply]
Thanks, this work for me.

**Reply by hirwablessing** [reply]
How do I get a response error status code then @TkDodo ?

**Reply by TkDodo** [reply]
> How do I get a response error status code then?

also depends on what you use to produce the promise, and is not part of react-queries's responsibilities. If you use `axios` for example, you can do `error.response.status`.

**Reply by hirwablessing** [reply]
I thought there might be a way react-query handles it. Thanks for the info!

**Reply by mikamokedo** [reply]
you save my life!

**Reply by imbishalgiri** [reply]
but typescript wont let you access error.response :(

**Reply by TkDodo** [reply]
@imbishalgiri see: https://tkdodo.eu/blog/react-query-and-type-script#what-about-error

**Reply by wonghin** [reply]
It works fine. Thank you

The solution would be
`error['response'].data`, so that there is no error in the editor (vscode)

**Reply by wonghin** [reply]
In Tanstack query, it would be
`onError:(error)=>{console.log(error['response'].data)}`

**Reply by onlyhasbi** [reply]
@wonghin thank you

**Reply by Queen-esther01** [reply]
This didnt work for me. Any idea why? This is what I get for error
```
Error: AxiosError: Request failed with status code 400
    at onResponseError (AxiosConfig.ts:156:10)
    at async uploadDocuments
```

**Reply by bringmetheaugust** [reply]
> This didnt work for me. Any idea why? This is what I get for error

This can happen if you use server actions or API routes in NextJS. An error that is initially an AxiosError instance is converted to a standard Error object when passed from the server to the client since JavaScript does not preserve object prototypes when serializing/deserializing

**Comment by olajoao** [comment]
For those struggling with Axios: You can configure a global error handler using Axios response interceptors:


```
axiosInstance.interceptors.response.use(
  function (response) {
    return response;
  },
  function (error) {
    if (error.response) {
      // handle the error here
      // example 
      toast.error(error.response.data.message);
    } else if (error.request) {
    // or here
      console.log("error.request", error.request);
    } else {
    // or here
    // example
      toast.error(error.message);
      console.log("error.message", error.message);
    }

    return Promise.reject(error);
  }
);

```

**Warning:** If you use a toast component, every single error are going to be shown

**Comment by yousifalraheem** [comment]
Is there a way to modify the error before it is stored in the `error` property? For example if I was running a request without useMutation, I would add an extra `.catch` that can "prettify" the error before it is ready to be consumed.

```js
new Promise((_, reject) => reject("Error!"))
  .catch((e) => {
      if (e instanceof Error) throw e;
      if (typeof e === 'string') throw new Error(e);
      throw new Error('Unknown error');
  })
```

My expectation is that I can just pass my error mutation in the `onError` and still get the error that I prepared in the `mutation.error`

Full example:

```js
const mutation = useMutation({
  mutationFn: async () => new Promise((_, reject) => reject("Error!")),
  onError: e => {
    if (e instanceof Error) throw e;
    if (typeof e === 'string') throw new Error(e);
    throw new Error('Unknown error');
  },
});

mutation.error.message // should be 'Unknown error', instead I get nothing, stuck on loading
```

**Reply by TkDodo** [reply]
`onError` is a way to “handle” the error you get back, not to `transform` errors. Just return the error you want from the `mutationFn`.

**Reply by yousifalraheem** [reply]
True, but handling an error can sometimes mean displaying it somewhere in the UI. With the current setup, I'll have to store the resolved error message in a state and use it. But that won't be as clear nor as convenient as having it stored in the mutation object itself.

By the way, this is how I also approach it if I was using `useAsync` function from `react-use`. It would be awesome if this is also supported in here.

**Reply by TkDodo** [reply]
I don’t understand what you want to achieve. You showed some code without mutations and that is exactly what you would do with a mutation too:

```
const mutation = useMutation({
  mutationFn: () => new Promise((_, reject) => reject("Error!"))
    .catch((e) => {
      if (e instanceof Error) throw e;
      if (typeof e === 'string') throw new Error(e);
      throw new Error('Unknown error');
  })
})
```

**Reply by TkDodo** [reply]
> instead I get nothing, stuck on loading

this shouldn’t happen though. Can you show a minimal reproduction please because looking at the code, this shouldn’t hang.
